sds011/sds011.cpp

```diff
@@ -0,0 +1,215 @@
+	termios tty;
+
+	if (tcgetattr(fh, &tty) < 0)
+		throw std::runtime_error(fmt::format("Failed to tcgetattr: {}", strerror(errno)));
+
+	cfsetospeed(&tty, (speed_t)speed);
+	cfsetispeed(&tty, (speed_t)speed);
+
+	tty.c_cflag |= (CLOCAL | CREAD); /* ignore modem controls */
+	tty.c_cflag &= ~CSIZE;
+	tty.c_cflag |= CS8;      /* 8-bit characters */
+	tty.c_cflag &= ~PARENB;  /* no parity bit */
+	tty.c_cflag &= ~CSTOPB;  /* only need 1 stop bit */
+	tty.c_cflag &= ~CRTSCTS; /* no hardware flowcontrol */
+
+	/* setup for non-canonical mode */
+	tty.c_iflag &= ~(IGNBRK | BRKINT | PARMRK | ISTRIP | INLCR | IGNCR | ICRNL | IXON);
+	tty.c_lflag &= ~(ECHO | ECHONL | ICANON | ISIG | IEXTEN);
+	tty.c_oflag &= ~OPOST;
+
+	/* fetch bytes as they become available */
+	tty.c_cc[VMIN] = 1;
+	tty.c_cc[VTIME] = 1;
+
+	if (tcsetattr(fh, TCSANOW, &tty) != 0)
+		throw std::runtime_error(fmt::format("Failed to tcsetattr: {}", strerror(errno)));
+
+	tty.c_cflag |= CS8;      /* 8-bit characters */
+	tty.c_cflag &= ~PARENB;  /* no parity bit */
+	tty.c_cflag &= ~CSTOPB;  /* on#pack-TEMPLATE%2cLISTly need 1 stop bit */
+	tty.c_cflag &= ~CRTSCTS; /* no hardware flowcontrol */
+
+	/* setup for non-canonical mode */
+	tty.c_iflag &= ~(IGNBRK | BRKINT | PARMRK | ISTRIP | INLCR | IGNCR | ICRNL | IXON);
+	tty.c_lflag &= ~(ECHO | ECHONL | ICANON | ISIG | IEXTEN);
+	tty.c_oflag &= ~OPOST;
+
+	/* fetch bytes as they become available */
+	tty.c_cc[VMIN] = 1;
+	tty.c_cc[VTIME] = 1;
+
+	if (tcsetattr(fh, TCSANOW, &tty) != 0)
+		throw std::runtime_error(fmt::format("Failed to tcsetattr: {}", strerror(errno)));
+}
+
+void set_blocking(int fh, int mcount) {
+	termios tty;
+
+	if (tcgetattr(fh, &tty) < 0)
+		throw std::runtime_error(fmt::format("Failed to tcgetattr: {}", strerror(errno)));
+
+	tty.c_cc[VMIN] = mcount ? 1 : 0;
+	tty.c_cc[VTIME] = 5; /* half second timer */
+
+	if (tcsetattr(fh, TCSANOW, &tty) < 0)
+		throw std::runtime_error(fmt::format("Failed to tcsetattr: {}", strerror(errno)));
+}
+};  // namespace
+
+sds011::sds011(const char* const path) : fh{open(path, O_RDWR | O_NOCTTY | O_SYNC)} {
+	if (fh < 0)
+		throw std::runtime_error(fmt::format("Failed to open '{}': {}", path, strerror(errno)));
+
+	try {
+		if (tcgetattr(fh, &tty_back) < 0)
+			throw std::runtime_error(fmt::format("Failed to tcgetattr: {}", strerror(errno)));
+
+		configure_interface(fh, B9600);
+		set_blocking(fh, 0);
+
+		/* There is a problem with flushing buffers on a serial USB that can
+		 * not be solved. The only thing one can try is to flush any buffers
+		 * after some delay:
+		 *
+		 * https://bugzilla.kernel.org/show_bug.cgi?id=5730
+		 * https://stackoverflow.com/questions/13013387/clearing-the-serial-ports-buffer
+		 */
+		usleep(10000);
+		tcflush(fh, TCIOFLUSH);
+	} catch (...) {
+		close(fh);
+		throw;
+	}
+}
+
+sds011::~sds011() {
+	if (tcsetattr(fh, TCSANOW, &tty_back) < 0)
+		fmt::print(stderr, "Failed to reset tcsetattr: {}", strerror(errno));
```

Comment:
exit(-1)?

Reply:
don't think it's a critical one; new tty reader/writer should set it own attrs anyways